ddtrace/profiling/collector/_traceback.pyx

```diff
@@ -1,22 +1,57 @@
+IF PY_MAJOR_VERSION > 3 or (PY_MAJOR_VERSION == 3 and PY_MINOR_VERSION >= 9):
+    from cpython cimport PyFrameObject
+    from cpython.object cimport PyObject
+    from cpython.ref cimport Py_XDECREF
+
+    cdef extern from "<Python.h>":
+        ctypedef struct PyCodeObject:
+            pass
+        PyObject* PyTuple_GetItem(PyObject *p, Py_ssize_t pos)
+        PyObject* PyDict_GetItem(PyObject *p, PyObject *key)
+        PyCodeObject* PyFrame_GetCode(PyFrameObject* frame)
+        PyFrameObject* PyFrame_GetBack(PyFrameObject* frame)
+
+        IF PY_MINOR_VERSION >= 11:
+            PyObject* PyFrame_GetLocals(PyFrameObject* frame)
+            int PyFrame_GetLineNumber(PyFrameObject* frame)
+            PyObject* PyCode_GetVarnames(PyCodeObject* co)
+
+
 cpdef _extract_class_name(frame):
     # type: (...) -> str
     """Extract class name from a frame, if possible.
 
     :param frame: The frame object.
     """
-    if frame.f_code.co_varnames:
-        argname = frame.f_code.co_varnames[0]
-        try:
-            value = frame.f_locals[argname]
-        except KeyError:
-            return ""
-        try:
-            if argname == "self":
-                return object.__getattribute__(type(value), "__name__")  # use type() and object.__getattribute__ to avoid side-effects
-            if argname == "cls":
-                return object.__getattribute__(value, "__name__")
-        except AttributeError:
-            return ""
+    # Python 3.11 moved PyFrameObject to internal C API and cannot be directly accessed from tstate
+    IF PY_MAJOR_VERSION > 3 or (PY_MAJOR_VERSION == 3 and PY_MINOR_VERSION >= 11):
+        code = PyFrame_GetCode(<PyFrameObject*> frame)
+        co_varnames = PyCode_GetVarnames(code)
+        if co_varnames:
+            argname = PyTuple_GetItem(co_varnames, 0)
+            try:
+                f_locals = PyFrame_GetLocals(<PyFrameObject*>frame)
+                value = <object>PyDict_GetItem(f_locals, argname)
+                Py_XDECREF(f_locals)
+                Py_XDECREF(argname)
+                Py_XDECREF(co_varnames)
```

Comment:
these should be handled by Cython for us ?

If you do want to manage these yourself, then you need to make sure they are called in ALL scenarios. e.g. if there is an exception raised then no XDECREF will get called. To handle better, you can predefine all of the `PyObject*` vars earlier and do a `try/finally` or make sure all branches will decref them.

Reply:
I'm hesitant to let Cython handle this for us, considering the source of the memory leak was due to a strong reference not being XDECREF'd. 
I'll go with your latter idea.